pkg/controllers/provisioning/scheduling/topology.go

```diff
@@ -18,149 +18,506 @@ import (
+	if err != nil {
+		return nil, fmt.Errorf("creating label selector: %w", err)
 	}
-	return topologyGroups
+	return &TopologyGroup{
+		Key:         cs.TopologyKey,
+		MaxSkew:     cs.MaxSkew,
+		Type:        TopologyTypeSpread,
+		rawSelector: cs.LabelSelector,
+		selector:    selector,
+		domains:     map[string]int32{},
+		namespaces:  sets.NewSet(namespace),
+		owners:      map[client.ObjectKey]struct{}{},
+	}, nil
 }
 
-func (t *Topology) computeCurrentTopology(ctx context.Context, constraints *v1alpha5.Constraints, topologyGroup *TopologyGroup) error {
-	switch topologyGroup.Constraint.TopologyKey {
-	case v1.LabelHostname:
-		return t.computeHostnameTopology(topologyGroup, constraints)
-	case v1.LabelTopologyZone:
-		return t.computeZonalTopology(ctx, constraints, topologyGroup)
-	case v1alpha5.LabelCapacityType:
-		return t.computeLabelCapacity(ctx, constraints, topologyGroup)
-	default:
-		return nil
+func (t *Topology) newForAffinity(term v1.PodAffinityTerm, namespaces sets.Set, topoType TopologyType) (*TopologyGroup, error) {
+	selector, err := metav1.LabelSelectorAsSelector(term.LabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("creating label selector: %w", err)
 	}
+
+	return &TopologyGroup{
+		Key:        term.TopologyKey,
+		MaxSkew:    math.MaxInt32,
+		Type:       topoType,
+		selector:   selector,
+		domains:    map[string]int32{},
+		namespaces: namespaces,
+		owners:     map[client.ObjectKey]struct{}{},
+	}, nil
 }
 
-// computeHostnameTopology for the topology group. Hostnames are guaranteed to
-// be unique when new nodes join the cluster. Nodes that join the cluster do not
-// contain any pods, so we can assume that the global minimum domain count for
-// `hostname` is 0. Thus, we can always improve topology skew (computed against
-// the global minimum) by adding pods to the cluster. We will generate
-// len(pods)/MaxSkew number of domains, to ensure that skew is not violated for
-// new instances.
-func (t *Topology) computeHostnameTopology(topologyGroup *TopologyGroup, constraints *v1alpha5.Constraints) error {
-	domains := []string{}
-	for i := 0; i < int(math.Ceil(float64(len(topologyGroup.Pods))/float64(topologyGroup.Constraint.MaxSkew))); i++ {
-		domains = append(domains, strings.ToLower(randomdata.Alphanumeric(8)))
+// trackInverseAntiAffinity is used to track topologies of inverse anti-affinities. Here the domains & counts track the
+// pods with the anti-affinity.
+func (t *Topology) trackInverseAntiAffinity(ctx context.Context, node *v1.Node, p *v1.Pod) error {
+	// We intentionally don't track inverse anti-affinity preferences.  We're not required to enforce them so it
+	// just adds complexity for very little value.
+	for _, term := range p.Spec.Affinity.PodAntiAffinity.RequiredDuringSchedulingIgnoredDuringExecution {
+		namespaces, err := t.buildNamespaceList(ctx, term.Namespaces, term.NamespaceSelector)
+		if err != nil {
+			return err
+		}
+		namespaces.Insert(p.Namespace)
```

Comment:
Is the pod's namespace considered if namespaceSelector is provided? I thought it was just a default.